components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt

```diff
@@ -0,0 +1,168 @@
+        const val CRASH_FILE_NAME = "glean_crash_counts"
+
+        // These keys correspond to the labels found for crashCount metric in metrics.yaml as well
+        // as the persisted crashes in the crash count file (see above comment)
+        const val UNCAUGHT_EXCEPTION_KEY = "uncaught_exception"
+        const val NATIVE_CODE_CRASH_KEY = "native_code_crash"
+    }
+
+    private val logger = Logger("glean/GleanCrashReporterService")
+
+    init {
+        run {
+            // We only want to record things on the main process because that is the only one in which
+            // Glean is properly initialized.  Checking to see if we are on the main process here will
+            // prevent the situation that arises because the integrating app's Application will be
+            // re-created when prompting to report the crash, and Glean is not initialized there since
+            // it's not technically the main process.
+            if (!context.isMainProcess()) {
+                logger.info("GleanCrashReporterService initialized off of main process")
+                return@run
+            }
+
+            if (!Glean.isInitialized()) {
+                logger.error("Glean is not initialized, crash reporter service cannot record metrics")
+                return@run
+            }
+
+            if (!checkFileConditions()) {
+                // checkFileConditions() internally logs error conditions
+                return@run
+            }
+
+            // Parse the persisted crashes
+            parseCrashFile()
+
+            // Clear persisted counts by deleting the file
+            file.delete()
+        }
+    }
+
+    /**
+     * Checks the file conditions to ensure it can be opened and read.
+     *
+     * @return True if the file exists and is able to be read, otherwise false
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun checkFileConditions(): Boolean {
+        return if (!file.exists()) {
+            // This is just an info line, as most of the time we hope there is no file which means
+            // there were no crashes
+            logger.info("No crashes to record, or file not found.")
+            false
+        } else if (!file.canRead()) {
+            logger.error("Cannot read file")
+            false
+        } else if (!file.isFile) {
+            logger.error("Expected file, but found directory")
+            false
+        } else {
+            true
+        }
+    }
+
+    /**
+     * Parses the crashes collected in the persisted crash file.  The format of this file is simple,
+     * each line may contain either [UNCAUGHT_EXCEPTION_KEY], or [NATIVE_CODE_CRASH_KEY] followed by
+     * a newline character.
+     *
+     * Example:
+     *
+     * <--Beginning of file-->
+     * uncaught_exception\n
+     * uncaught_exception\n
+     * native_code_crash\n
+     * uncaught_exception\n
+     * <--End of file-->
```

Comment:
Nothing blocking here, but what would be the strategy to expand this in the future? Switch to a different file with new semantics? Or introduce some versioning to know which format this is in?

Reply:
We are moving to Reference Browser as soon as this lands!  As far as expanding on this, I think we would probably switch to a different file with different semantics, as there will likely be more metrics than just the counts that we want to capture.  That's one of the next big things on my radar is to expand crash metrics.